server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3831,14 +3831,17 @@ public void afterRefresh(boolean didRefresh) {
      * @param listener the listener to invoke once the pending refresh location is visible. The listener will be called with
      *                 <code>true</code> if the listener was registered to wait for a refresh.
      */
-    public final void awaitShardSearchActive(Consumer<Boolean> listener) {
+    public final void makeShardSearchActive(Consumer<Boolean> listener) {
         markSearcherAccessed(); // move the shard into non-search idle
         final Translog.Location location = pendingRefreshLocation.get();
         if (location != null) {
             addRefreshListener(location, (result) -> {
                 pendingRefreshLocation.compareAndSet(location, null);
                 listener.accept(true);
             });
+            // TODO: maybe just invoke getEngine().maybeRefresh(...) here?
+            // (schedule refresh does a few more things that I don't is necessary here?)
+            scheduledRefresh();
```

Comment:
I think it will work fine, but one problem with `maybeRefresh` is that it will not provoke a refresh if there is an ongoing one. But there is no certainty that another ongoing refresh will contain the location. However, such a refresh would have to be an explicit one. I find the risk small and regardless an improvement over current state.

I think we can switch to `maybeRefresh` instead, though I'd like to check that `pendingRefreshLocation` is still `location` then, i.e.:

```suggestion
            if (pendingRefreshLocation.get() == location) {
               getEngine().maybeRefresh();
           }
```

both to handle racy cases and the case where there are no more refresh listener slots.